### ISSUE TITLE
potential issue with tracker and database usage

### ISSUE THREAD

**Issue by ldpr** [torrust/torrust-tracker#997]
sorry for the rather vague title, i struggled to think of something better.

it's not uncommon for us to be doing ~100 requests per second and we're still relatively small. (this is with a 30 minute announce time also). i can't really test this in production _yet_ but i'd be happy to if you guys think it will be helpful.

regardless, that'll end up with a _lot_ of reads/writes to the database which i don't think is necessary? Or rather, if you'd like a database integration an in-memory solution such as valkey/redis would be better suited (IMO). i should note too that sqlite is performing brilliantly for us under the above load so in our case, we'd just stick with that rather than linking it to mysql. torrust as it stands is extremely resource friendly!

i'm not sure it's of all that much benefit for the tracker to have a stateful database really. overhead issues aside, if it was stateless (including redis), the worst that can happen is clients will just re-announce and pickup where they left off and in most cases, users won't even notice. it's also very rare that the tracker would need to be restarted (or indeed torrust) as they're incredibly stable. for example, it's noticeable when i restart the tracker as you can see seeds drop off on the site but after 30 minutes everyone has announced it's back up to normal like nothing happened.

what do you think?
edit; after doing a bit more research 100 writes per second is possible with traditional hard drives https://dba.stackexchange.com/questions/315314/approximately-how-many-reads-writes-per-secons-can-mysql-mariadb-tolerate and SSDs/NVMEs can scale far beyond that. there isn't any info regarding the CPU/RAM usage for this though. 

TL:DR I think mysql is a square-peg in a round-hole for the _tracker_ specifically.

**Comment by josecelano** [comment]
Hi @ldpr thank you for your feedback.

To be honest, I haven't checked MySQL's performance. I want to start using MySQL on the [live demo](https://index.torrust-demo.com/torrents). Adding some indexes might improve performance.

Regarding your second point. The tracker does **NOT** use the database for the core tracker services. All information about swarms is only stored in memory.

The database is only used for these three features:

1. Torrents stats (for now, only number of downloads).
2. Whitelist.
3. User's keys (only when the tracker is running in private mode)

Regarding **torrents stats**, you can disable persisting them with the following:

```toml
[core.tracker_policy]
persistent_torrent_completed_stat = false
```

Regarding the **whitelist** feature, you can also disable it:

```toml
[core]
listed = false
```

And finally, regarding **user's keys**, that only affects when the tracker is private:

```toml
[core]
private = true
```

However, keys are loaded when the tracker starts. The database is only used when the application adds or deletes a key, but not for normal BitTorrent requests (announce/scrape).

We are using [custom in-memory structs for swarm data](https://github.com/torrust/torrust-tracker/tree/develop/packages/torrent-repository). We called it [torrent repository](https://github.com/torrust/torrust-tracker/tree/develop/packages/torrent-repository).

I've been considering using an [in-memory key-value database](https://github.com/torrust/torrust-tracker/discussions/774#discussioncomment-9000983) but we have not made any test yet. See https://github.com/torrust/torrust-tracker/discussions/774#discussioncomment-9000983

**Comment by ldpr** [comment]
Hi Jose!

Thank you! That makes a lot more sense than I'd assumed and that's my fault. Wonderful, sounds fantastic. I'll go ahead and close this out then, I have some more reading to do. Thanks again :)
